### PR TITLE
GEOMESA-2177 Reduce developer friction using Spark JTS

### PR DESCRIPTION
**Pull request by metasim**
See https://geomesa.atlassian.net/browse/GEOMESA-2177 for overview and rationale.

## Tasks

* [x] Update documentation

### CODE REVIEW

**Comment by metasim** [comment]
Note: This PR will be easier to review once 2173 is merged and this can be rebased against master without 26a0a55 in the history.

**Comment by elahrvivaz** [comment]
@metasim I've merged #1853, you can rebase this now.
FYI, when responding to comments you can just add additional commits - that makes it easier to see what has changed. We always do a merge squash at the end, anyway.

**Comment by metasim** [comment]
@elahrvivaz Thanks for merging that! I was worried I'd have to wait all day for the Travis build!

**Comment by elahrvivaz** [comment]
if it's broken, you can fix it here ;)

**Comment by elahrvivaz** [comment]
@metasim did you rebase this yet? if so, the diff seems confused I think... I was waiting to review until you rebased, just want to make sure you're not waiting on me. thanks,

**Comment by elahrvivaz** [comment]
@metasim ok thanks. so the diff (files changed) looks right to you?

**Comment by metasim** [comment]
I suggest reviewing 745d188 first, to get a sense on where things were moved (note IntelliJ didn't always do the `git mv` right, so some files look like they were deleted). Then review the second two commits to see how some classes/functions were pulled out into their own files.

The final result is that you can do this:
```
import  org.locationtech.geomesa.spark.jts._
...
val spark: SparkSession = _
spark.withJTS // or if you don't like method syntax: `initJTS(spark)`
```

**Comment by metasim** [comment]
Bird's eye view of the resulting directory structure:

```
geomesa-spark-jts
└── src
    ├── main
    │   └── scala
    │       └── org
    │           ├── apache
    │           │   └── spark
    │           │       └── sql
    │           │           └── jts
    │           └── locationtech
    │               └── geomesa
    │                   └── spark
    │                       └── jts
    │                           ├── encoders
    │                           ├── rules
    │                           ├── udaf
    │                           ├── udf
    │                           └── util
    └── test
        ├── resources
        └── scala
            └── org
                └── locationtech
                    └── geomesa
                        └── spark
                            └── jts
                                └── udf
```

File: geomesa-spark/geomesa-spark-jts/src/main/scala/org/locationtech/geomesa/spark/jts/package.scala [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+   * enrichment methods on [[SQLContext]] or [[SparkSession]] can be used instead.
+   */
+  def initJTS(sqlContext: SQLContext): Unit = {
+    org.apache.spark.sql.jts.registerTypes()
+    udf.registerFunctions(sqlContext)
+    rules.registerOptimizations(sqlContext)
+  }
+
+  /** Enrichment over [[SQLContext]] to add `withJTS` "literate" method. */
+  implicit class SQLContextWithJTS(sqlContext: SQLContext) {
```
Comment (elahrvivaz) [inline geomesa-spark/geomesa-spark-jts/src/main/scala/org/locationtech/geomesa/spark/jts/package.scala]: can we make this `implicit class SQLContextWithJTS(val sqlContext: SQLContext) extends AnyVal` - that way we make it a value class and can avoid allocating a new class instance: https://docs.scala-lang.org/overviews/core/value-classes.html

File: geomesa-spark/geomesa-spark-jts/src/main/scala/org/locationtech/geomesa/spark/jts/package.scala [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+    def withJTS: SQLContext = {
+      initJTS(sqlContext)
+      sqlContext
+    }
+  }
+
+  /** Enrichment over [[SparkSession]] to add `withJTS` "literate" method. */
+  implicit class SparkSessionWithJTS(spark: SparkSession) {
```
Comment (elahrvivaz) [inline geomesa-spark/geomesa-spark-jts/src/main/scala/org/locationtech/geomesa/spark/jts/package.scala]: could also make this a value class

File: geomesa-spark/geomesa-spark-jts/src/main/scala/org/locationtech/geomesa/spark/jts/udf/GeometricConstructorFunctions.scala [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+    ST_MPointFromText -> "st_mPointFromText",
+    ST_MPolyFromText -> "st_mPolyFromText",
+    ST_Point -> "st_point",
+    ST_PointFromText -> "st_pointFromText",
+    ST_PointFromWKB -> "st_pointFromWKB",
+    ST_Polygon -> "st_polygon",
+    ST_PolygonFromText  -> "st_polygonFromText"
+  )
+
+
```
Comment (elahrvivaz) [inline geomesa-spark/geomesa-spark-jts/src/main/scala/org/locationtech/geomesa/spark/jts/udf/GeometricConstructorFunctions.scala]: [format] remove extra newlines

File: geomesa-spark/geomesa-spark-jts/src/main/scala/org/locationtech/geomesa/spark/jts/util/ObjectPoolUtils.scala [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+ * which accompanies this distribution and is available at
+ * http://www.opensource.org/licenses/apache2.0.php.
+ ***********************************************************************/
+
+package org.locationtech.geomesa.spark.jts.util
+
+import org.apache.commons.pool.impl.GenericObjectPool
+
+// NB: This is a duplicate of a file in geomesa-utils
+trait ObjectPoolUtils[A] {
```
Comment (elahrvivaz) [inline geomesa-spark/geomesa-spark-jts/src/main/scala/org/locationtech/geomesa/spark/jts/util/ObjectPoolUtils.scala]: can we actually just get rid of this class? I realize that in WKTUtils we switched to thread locals.

File: geomesa-spark/geomesa-spark-jts/src/main/scala/org/locationtech/geomesa/spark/jts/package.scala (hunk 1 above)
Comment (metasim) [reply geomesa-spark/geomesa-spark-jts/src/main/scala/org/locationtech/geomesa/spark/jts/package.scala]: +1

File: geomesa-spark/geomesa-spark-jts/src/main/scala/org/locationtech/geomesa/spark/jts/util/ObjectPoolUtils.scala (hunk 4 above)
Comment (metasim) [reply geomesa-spark/geomesa-spark-jts/src/main/scala/org/locationtech/geomesa/spark/jts/util/ObjectPoolUtils.scala]: I'll remove and see what happens.

**Comment by elahrvivaz** [comment]
@metasim we can remove the WIP tag now right? Do we need to update documentation like in your TODO, or is this ready to merge?

**Comment by metasim** [comment]
@elahrvivaz Ready to merge.

File: docs/user/spark/spark_jts.rst [hunk 5]
```diff
@@ -26,53 +26,60 @@ Example
...
+    val df = Seq(point).toDF("point")
 
 Configuration
 ^^^^^^^^^^^^^
 
-To enable this behavior, you must pass the  ``SQLContext`` of the ``SparkSession`` to
-``JTSTypes.init``. This will register the UDFs and UDTs as well as some catalyst
-optimizations for these operations.
+To enable this behavior, import `org.locationtech.geomesa.spark.jts._`, create a
+``SparkSession` and call `.withJTS` on it. This will register the UDFs and UDTs as
```
Comment (aheyne) [inline docs/user/spark/spark_jts.rst]: Back tics got messed up here a bit, need double for each.

File: docs/user/spark/spark_jts.rst [hunk 6]
```diff
@@ -26,53 +26,60 @@ Example
...
 
     import org.apache.spark.sql.SparkSession
     import org.apache.spark.sql.SQLContext
-    import org.apache.spark.sql.jts.JTSTypes
+    import org.locationtech.geomesa.spark.jts._
 
     val spark: SparkSession = SparkSession.builder() // ... initialize spark session
-    val sc: SQLContext = spark.sqlContext
-    JTSTypes.init(sc)
+    spark.withJTS
```
Comment (aheyne) [inline docs/user/spark/spark_jts.rst]: Nice

File: geomesa-spark/geomesa-spark-jts/src/main/scala/org/locationtech/geomesa/spark/jts/encoders/SpatialEncoders.scala [hunk 7]
```diff
@@ -0,0 +1,25 @@
+/***********************************************************************
```
Comment (aheyne) [inline geomesa-spark/geomesa-spark-jts/src/main/scala/org/locationtech/geomesa/spark/jts/encoders/SpatialEncoders.scala]: Are you interested in a adding a header?

File: geomesa-spark/geomesa-spark-jts/src/main/scala/org/locationtech/geomesa/spark/jts/rules/GeometryLiteral.scala [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+
+import com.vividsolutions.jts.geom.Geometry
+import org.apache.spark.sql.catalyst.InternalRow
+import org.apache.spark.sql.catalyst.expressions.LeafExpression
+import org.apache.spark.sql.catalyst.expressions.codegen.CodegenFallback
+import org.apache.spark.sql.jts.JTSTypes.GeometryTypeInstance
+import org.apache.spark.sql.types.DataType
+
+/**
+ * Catalyst AST expression used during rule rewriting to capture extract geometry literal values
```
Comment (aheyne) [inline geomesa-spark/geomesa-spark-jts/src/main/scala/org/locationtech/geomesa/spark/jts/rules/GeometryLiteral.scala]: Confusing sentence, typo I guess (capture/extract)?

**Pull request merged**